### PR TITLE
refactor!: change `ArrayLength` value from `uint256` to `uint128`

### PR DESCRIPTION
**Pull request by b00ste**
#### ⚠️ BREAKING CHANGES

This PR update how the array length for data of keyType `Array` is encoded.

Since it was changed and reduced from `uint256` to `uint128`.

See --> https://github.com/lukso-network/LIPs/pull/180

### CODE REVIEW

**Comment by b00ste** [comment]
This PR is on hold, focusing on other priorities right now.

File: src/lib/utils.ts
Comment (kalote) [inline src/lib/utils.ts:265]: So, you're not using `ArrayLength` type here and decided to move to a simpler approach?

File: src/lib/encoder.ts
Comment (Hugoo) [inline src/lib/encoder.ts:332]: You can't do like this?

```suggestion
  uint128: {
    encode: (value: string | number) => abiCoder.encodeParameter('uint128', value),
    decode: (value: string) => abiCoder.decodeParameter('uint128', value),
  },
```

File: src/lib/encoder.ts
Comment (Hugoo) [inline src/lib/encoder.ts:416]: Then let's remove this.

It used to work at a time with uint256.
At no point it time it has or will exist with utin128 as it is already deprecated. I suggest to remove it totally.

File: src/lib/utils.ts
Comment (Hugoo) [inline src/lib/utils.ts:266]: What is the reason for not using the `encodeKeyValue`?

`leftPad(value.length, 32)` is duplicated code with https://github.com/ERC725Alliance/erc725.js/pull/287/files#diff-5927840c3323c66acd20762ba79459d6e5fe967c32d6d2a4be8751fb589c0c93R330

We want only one place in the code that does the encoding, not to sprint it everywhere in the code no?

File: test/mockSchema.ts
Comment (Hugoo) [inline test/mockSchema.ts:229]: Here you have kept the abiCode.encodeParemeter, nice. Why didn't you update it the same way earlier? cf. the other comment:

https://github.com/ERC725Alliance/erc725.js/pull/287/files#diff-5927840c3323c66acd20762ba79459d6e5fe967c32d6d2a4be8751fb589c0c93R330

**Review by Hugoo — COMMENTED** [review]
lets clear some stuff before merging - cf comments

File: src/lib/encoder.ts
Comment (CJ42) [reply src/lib/encoder.ts:332]: No this does not work work because  it uses ABI encoding.

ABI encoding will always pas to 32 bytes, even if the type is smaller.

See the Solidity docs:

https://docs.soliditylang.org/en/v0.8.20/abi-spec.html

image

File: src/lib/utils.ts
Comment (CJ42) [reply src/lib/utils.ts:266]: I think because it was encoding as a valueContent "Number" and the "Number" valueContent in the code uses uint256 as the valueType to encode.

File: test/mockSchema.ts
Comment (CJ42) [reply test/mockSchema.ts:229]: I am actually not sure about this for the mockSchemas as it is not clear to me.

Here the same reason as before. Because ABI encoding a uint128 still result in a 32 bytes long value.

But we want it to be 16 bytes long.

File: src/lib/utils.ts
Comment (CJ42) [reply src/lib/utils.ts:265]: No I did not use this because this valueContent is not defined in LSP2.

I guess we might want to deprecate it.

See this discussion comments: https://github.com/ERC725Alliance/erc725.js/pull/287#discussion_r1193203269

File: src/lib/utils.ts
Comment (CJ42) [reply src/lib/utils.ts:266]: @Hugoo here, after debugging with @kalote we found that `encodeKeyValue` ends up encoding the value as a `Number`, which itself encodes as a `uint256` (and we want to encode Array length as `uint128`.

https://github.com/ERC725Alliance/erc725.js/blob/9e075c4318e0d0055b52b3b70c4ad50ac06c1412/src/lib/encoder.ts#L413-L428

There is a better way, using `encodeValueType` function (already used in this file).

File: test/mockSchema.ts
Comment (CJ42) [reply test/mockSchema.ts]: Ok I was misunderstanding and getting things mixed up. 
I have updated the `mockSchemas` to use `leftPad` the same way as it is done in the library for Array length

File: src/lib/encoder.ts
Comment (Hugoo) [reply src/lib/encoder.ts:332]: Ok I see, nice one, ty for the explanation

**Review by kalote — APPROVED** [review]

**Review by magalimorin18 — APPROVED** [review]

**Comment by Hugoo** [comment]
@b00ste is it still on hold as described here?
https://github.com/ERC725Alliance/erc725.js/pull/287#issuecomment-1479056978

**Comment by CJ42** [comment]
@Hugoo no it

> @b00ste is it still on hold as described here? [#287 (comment)](https://github.com/ERC725Alliance/erc725.js/pull/287#issuecomment-1479056978)

@Hugoo no it's not on hold anymore. This is ready to be merged once approved.

**Review by CallumGrindle — APPROVED** [review]

File: src/lib/encoder.ts
Comment (Hugoo) [inline src/lib/encoder.ts:330]: what happens if i pass "HelloMyFriend" as an input? It will encode as:

`0x000....000HelloMyFriend` which is not hex and something will fail very hard later in the process no?

What if input is `123456789` ? Will it encode as `0x000...0000123456789` ? Shouldn't it be hex (not decimal) ??

Do we have checks that prevent having corrupted values?

File: src/lib/utils.test.ts
Comment (Hugoo) [inline src/lib/utils.test.ts:513]: Can you ad test with values that are wrong? And expect it to throw?

Here we are in a very very happy path.

What if input is "hello world" ? What if input is "11" ?
Do we get `0x00...000011` ? Or should we get `0x00...0000C` ??

Add more test cases please

**Review by Hugoo — CHANGES_REQUESTED** [review]
i have a bad feeling with this encoding. I think it misses a lot of things (if number if > 9, if input is a string?). I feel like we need more validation here. I think the `abiCoder` was doing it for us, but now, you just add 000 on one side :/

File: src/lib/encoder.ts
Comment (Hugoo) [inline src/lib/encoder.ts:345]: I think you can write the functions here directly instead of making the functions hexToUint128 etc.

File: src/lib/encoder.ts
Comment (Hugoo) [inline src/lib/encoder.ts:320]:
```suggestion
    throw new Error(`Can't convert hex value ${value} to uint128. Too many bytes. ${value.length - 2 / 2} > 16`);
```

File: src/lib/encoder.ts
Comment (Hugoo) [inline src/lib/encoder.ts:318]:
```suggestion
  if (!isHex(value)) throw new Error(`Can't convert ${value} to uint128, value is not hex.`);
```

File: src/lib/utils.test.ts
Comment (Hugoo) [inline src/lib/utils.test.ts:569]: this is copy/pasted from line 525 to 535, put these in a variable and reuse this variable at both places rather than copy/pasting.

**Review by Hugoo — APPROVED** [review]

File: src/lib/encoder.ts
Comment (CJ42) [reply src/lib/encoder.ts:345]: I was about to put a review comment for the same thing.

We can probably extract to a function later, we will have to support multiple uintN

File: src/lib/encoder.ts
Comment (CJ42) [inline src/lib/encoder.ts:438]: Why converting it to a string? I think we should decode uintN data as number, not string.

Would suggest to remove this.

@Hugoo @b00ste

Comment (b00ste) [reply src/lib/encoder.ts:438]: It was converted to string before, instead of having `'' + hexToNumber(..)` i used the function `toString()`.

Comment (CJ42) [reply src/lib/encoder.ts:438]: I removed this behaviour in this PR afterwards.

So I guess we can already replace that here.

https://github.com/ERC725Alliance/erc725.js/pull/289

Comment (b00ste) [reply src/lib/encoder.ts:437]: Let's replace it in your next PR, because removing it breaks some tests

File: src/lib/encoder.ts
Comment (CJ42) [inline src/lib/encoder.ts:334]: Can we put curly brackets for `if` statements, so to keep it consistent with the rest of the code.

Like here:

image

**Review by CJ42 — CHANGES_REQUESTED** [review]

**Review by CJ42 — APPROVED** [review]

File: src/lib/encoder.ts
Comment (Hugoo) [inline src/lib/encoder.ts:44]: Let's not add this dep: fixed here: https://github.com/ERC725Alliance/erc725.js/pull/304

**Pull request merged**
